### PR TITLE
[RFC] path: check if directories are readable when listing files

### PR DESCRIPTION
**Pull request by fmoralesc**
Fixes https://github.com/neovim/neovim/issues/3164 and https://github.com/neovim/neovim/issues/3194

### CODE REVIEW

File: src/nvim/path.c [hunk 1]
```diff
@@ -612,8 +612,9 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
-      || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
+  if (os_file_is_readable(*buf == NUL ? "." : (char *)buf)
+      && (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
+          || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf))) {
```
Comment (justinmk) [inline src/nvim/path.c]: `(char_u *)buf` isn't needed is it?

Comment (fmoralesc) [inline src/nvim/path.c]: :+1: Give me a sec.

**Comment by Pyrohh** [comment]
Heh, this fixes the issue you mentioned for me, but now an assertion triggers for me upon exiting:
```
; ./build/bin/nvim -u NORC /root/
assertion "!(handle->flags & UV_CLOSED)" failed: file "/home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c", line 210, function "uv__finish_close"
abort--core dumped
;
```

I'll try updating to a newer libuv, maybe that will magically fix it :-)

**Comment by justinmk** [comment]
Not the same, but a similar report: https://github.com/neovim/neovim/issues/2499

There have been other cases recently as well.

**Comment by justinmk** [comment]
@Pyrohh could you get the full stack trace? Maybe this could help us fix some long-standing issues.

**Comment by fmoralesc** [comment]
I'm also using 1.7.3. Are you on *BSD, @Pyrohh? Maybe it's something platform-specific.

**Comment by Pyrohh** [comment]
@fmoralesc Yes, OpenBSD; maybe something is ifdef'd for it. Have you verified you're building with assertions (i.e., the Build Type is not `Release`)?

**Comment by fmoralesc** [comment]
Yes, I'm testing on the `Debug` build, and I can't reproduce.

**Comment by Pyrohh** [comment]
Just to be extra clear, `./build/bin/nvim -u NONE /root/` wasn't triggering anything, presumably because netrw isn't being loaded; `./build/bin/nvim /root/` had the same effect for me because this is at the top of my nvimrc:
```vim
" Disable some plugins
let g:loaded_netrwPlugin = 1
let g:loaded_vimballPlugin = 1
```

However, I was able to trigger the assertion with `./build/bin/nvim -u NORC /root/`.

**Comment by fmoralesc** [comment]
Just re-checked, and it acts the same with `-u NORC` and `-u NONE` here, except for netrw being loaded. What are the permissions of `/root` on OpenBSD? I have `0750/drwxr-x---` here (Arch Linux).

**Comment by Pyrohh** [comment]
`drwx------` for me.

**Comment by Pyrohh** [comment]
@justinmk here's a short backtrace:
```
#0  0x00000f1f19e661ea in kill () at <stdin>:2
#1  0x00000f1f19ea00b9 in abort () at /usr/src/lib/libc/stdlib/abort.c:53
#2  0x00000f1f19e1aa24 in __assert2 (file=<optimized out>, line=<optimized out>, func=<optimized out>, 
    failedexpr=<optimized out>) at /usr/src/lib/libc/gen/assert.c:52
#3  0x00000f1d0593ab52 in uv__finish_close (handle=0xf1f6b7d04c8)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:224
#4  uv__run_closing_handles (loop=<optimized out>)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:273
#5  uv_run (loop=0xf1d05de7120 <loop>, mode=UV_RUN_DEFAULT)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:343
#6  0x00000f1d059335a7 in loop_close (loop=0xf1d05de7120 <loop>) at ../src/nvim/event/loop.c:79
#7  0x00000f1d05808d4d in event_teardown () at ../src/nvim/main.c:172
#8  0x00000f1d0587019e in mch_exit (r=0) at ../src/nvim/os_unix.c:180
#9  0x00000f1d05809fc9 in getout (exitval=0) at ../src/nvim/main.c:831
#10 0x00000f1d057b81ea in ex_quit (eap=0x7f7fffff66e0) at ../src/nvim/ex_docmd.c:5635
#11 0x00000f1d057af928 in do_one_cmd (cmdlinep=0x7f7fffff6e58, sourcing=1, cstack=0x7f7fffff68c0, fgetline=0x0, 
    cookie=0x0) at ../src/nvim/ex_docmd.c:2195
#12 0x00000f1d057ab717 in do_cmdline (cmdline=0xf1d05a79dfd "q!", fgetline=0x0, cookie=0x0, flags=11)
    at ../src/nvim/ex_docmd.c:621
#13 0x00000f1d057aae2a in do_cmdline_cmd (cmd=0xf1d05a79dfd "q!") at ../src/nvim/ex_docmd.c:293
#14 0x00000f1d0584a318 in nv_Zet (cap=0x7f7fffff6f60) at ../src/nvim/normal.c:4181
#15 0x00000f1d05842767 in normal_cmd (oap=0x7f7fffff7030, toplevel=true) at ../src/nvim/normal.c:912
#16 0x00000f1d05809d0c in main_loop (cmdwin=0, noexmode=0) at ../src/nvim/main.c:747
#17 0x00000f1d058096da in main (argc=4, argv=0x7f7fffff72a8) at ../src/nvim/main.c:534
```

and `bt full`:

```
#0  0x00000f1f19e661ea in kill () at <stdin>:2
No locals.
#1  0x00000f1f19ea00b9 in abort () at /usr/src/lib/libc/stdlib/abort.c:53
        mask = 4294967263
#2  0x00000f1f19e1aa24 in __assert2 (file=<optimized out>, line=<optimized out>, func=<optimized out>, 
    failedexpr=<optimized out>) at /usr/src/lib/libc/gen/assert.c:52
No locals.
#3  0x00000f1d0593ab52 in uv__finish_close (handle=0xf1f6b7d04c8)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:224
        __func__ = "uv__finish_close"
#4  uv__run_closing_handles (loop=<optimized out>)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:273
No locals.
#5  uv_run (loop=0xf1d05de7120 <loop>, mode=UV_RUN_DEFAULT)
    at /home/michael/src/neovim/.deps/build/src/libuv/src/unix/core.c:343
        timeout = <optimized out>
        r = <optimized out>
#6  0x00000f1d059335a7 in loop_close (loop=0xf1d05de7120 <loop>) at ../src/nvim/event/loop.c:79
No locals.
#7  0x00000f1d05808d4d in event_teardown () at ../src/nvim/main.c:172
No locals.
#8  0x00000f1d0587019e in mch_exit (r=0) at ../src/nvim/os_unix.c:180
No locals.
#9  0x00000f1d05809fc9 in getout (exitval=0) at ../src/nvim/main.c:831
        tp = 0x0
        next_tp = 0x0
#10 0x00000f1d057b81ea in ex_quit (eap=0x7f7fffff66e0) at ../src/nvim/ex_docmd.c:5635
        wp = 0xf1f66981000
#11 0x00000f1d057af928 in do_one_cmd (cmdlinep=0x7f7fffff6e58, sourcing=1, cstack=0x7f7fffff68c0, fgetline=0x0, 
    cookie=0x0) at ../src/nvim/ex_docmd.c:2195
        p = 0xf1fad85cd42 ""
        lnum = 2147483647
        n = 0
        errormsg = 0x0
        ea = {arg = 0xf1fad85cd42 "", nextcmd = 0x0, cmd = 0xf1fad85cd40 "q!", cmdlinep = 0x7f7fffff6e58, 
          cmdidx = CMD_quit, argt = 1066243, skip = 0, forceit = 1, addr_count = 0, line1 = 1, line2 = 1, 
          addr_type = 1, flags = 0, do_ecmd_cmd = 0x0, do_ecmd_lnum = 0, append = 0, usefilter = 0, amount = 0, 
          regname = 0, force_bin = 0, read_edit = 0, force_ff = 0, force_enc = 0, bad_char = 0, useridx = 0, 
          errmsg = 0x0, getline = 0x0, cookie = 0x0, cstack = 0x7f7fffff68c0}
        verbose_save = -1
        save_msg_scroll = 0
        save_msg_silent = -1
        did_esilent = 0
        did_sandbox = 0
        save_cmdmod = {hide = 0, split = 0, tab = 0, confirm = 0, keepalt = 0, keepmarks = 0, keepjumps = 0, 
          lockmarks = 0, keeppatterns = 0, noswapfile = false, save_ei = 0x0}
        ni = 0
        cmd = 0xf1fad85cd40 "q!"
#12 0x00000f1d057ab717 in do_cmdline (cmdline=0xf1d05a79dfd "q!", fgetline=0x0, cookie=0x0, flags=11)
    at ../src/nvim/ex_docmd.c:621
        next_cmdline = 0xf1fad85cd40 "q!"
        cmdline_copy = 0xf1fad85cd40 "q!"
        used_getline = 0
        msg_didout_before_start = 0
        count = 1
        did_inc = 0
        retval = 1
        cstack = {cs_flags = {0 <repeats 50 times>}, cs_pending = '\000' <repeats 49 times>, cs_pend = {csp_rv = {
              0x0 <repeats 33 times>, 0x4, 0xf1d0594733f <uv__hrtime+31>, 0x1bb5, 0xf1d05943fe8 <uv__run_timers+72>, 
              0x7f7fffff6b80, 0xee171ecc43237775, 0xf1d05de7120 <loop>, 0x0, 0xf1d05de7120 <loop>, 0xf1f82226cc0, 
              0x7f7fffff6ad0, 0xf1d0593aaeb <uv_run+667>, 0x7f7fffff6b20, 0xf1d059354ec <queue_process_events+108>, 
              0x0, 0xf1f82226cc0, 0x0}, csp_ex = {0x0 <repeats 33 times>, 0x4, 0xf1d0594733f <uv__hrtime+31>, 0x1bb5, 
              0xf1d05943fe8 <uv__run_timers+72>, 0x7f7fffff6b80, 0xee171ecc43237775, 0xf1d05de7120 <loop>, 0x0, 
              0xf1d05de7120 <loop>, 0xf1f82226cc0, 0x7f7fffff6ad0, 0xf1d0593aaeb <uv_run+667>, 0x7f7fffff6b20, 
              0xf1d059354ec <queue_process_events+108>, 0x0, 0xf1f82226cc0, 0x0}}, cs_forinfo = {0xee171ecc43237775, 
            0xf1fbcd26070, 0x7f7fffff72d0, 0x7f7fffff72d0, 0xf1ff1743000, 0x1bb5, 0xf1ff1743000, 0x7f7fffff6b50, 
            0xf1d0591e080 <input_ready+33>, 0x7f7fffff6b70, 0xf1d05880395 <rbuffer_consumed+161>, 0x1bcd16000, 0x0, 
            0x7f7fffff6bb8, 0xf1ff1743000, 0xf1fbcd16048, 0x7f7fffff72d0, 0x7f7fffff6bd0, 
            0xf1d0588059b <rbuffer_read+176>, 0x1, 0x54, 0xf1d05ddfee8 <typebuf_init+8>, 0xf1ff1743000, 
            0x7f7fffff6c10, 0xf1d057f071c <fix_input_buffer+23>, 0x0, 0x100000000, 0x55, 0xee171ecc43237775, 
            0x7f7fffff6c10, 0xf1d0591d75d <os_inchar+272>, 0x100000000, 0xf1d05ddfee7 <typebuf_init+7>, 0x55ffffffff, 
            0x800000000, 0x100000001, 0xf1d05ddfee6 <typebuf_init+6>, 0x0, 0x0, 0x1bb5, 0xee171ecc43237775, 0x0, 0x1, 
            0xf1d05de7120 <loop>, 0x7f7fffff72a8, 0x7f7fffff6c80, 0xf1f82226cc0, 0xf1d05de7120 <loop>, 
            0xf1d0593aaeb <uv_run+667>, 0x7f7fffff6cd0}, cs_line = {93541612, 3869, 98464032, 3869, -2111673152, 
            3871, -37648, 32639, 1126397813, -300474676, -1127129016, 3871, -36144, 0, -37632, 32639, -36184, 32639, 
            -37664, 32639, 92199032, 3869, -37632, 32639, 93533475, 3869, -37600, 32639, 92198897, 3869, -37600, 1, 
            98434791, 3869, 98434792, 3869, 0, 1, 81, 0, 1126397813, -300474676, -37280, 32639, 92206116, 3869, 90, 
            0, 1126397813, 81}, cs_idx = -1, cs_looplevel = 0, cs_trylevel = 0, cs_emsg_silent_list = 0x0, 
          cs_lflags = 0 '\000'}
        lines_ga = {ga_len = 0, ga_maxlen = 0, ga_itemsize = 16, ga_growsize = 10, ga_data = 0x0}
        current_line = 0
        fname = 0x0
        breakpoint = 0x0
        dbg_tick = 0x0
        debug_saved = {trylevel = 1721249792, force_abort = 3871, caught_stack = 0x80000005a, 
          vv_exception = 0xffffffffffffffff <error: Cannot access memory at address 0xffffffffffffffff>, 
          vv_throwpoint = 0xffffffffffffffff <error: Cannot access memory at address 0xffffffffffffffff>, 
          did_emsg = -65534, got_int = 32639, did_throw = 439082156, need_rethrow = 1, check_cstack = -1, 
          current_exception = 0xffffffffffffffff}
        initial_trylevel = 0
        saved_msg_list = 0x0
        private_msg_list = 0x0
        cmd_getline = 0x0
        cmd_cookie = 0x0
        cmd_loop_cookie = {lines_gap = 0xf1fad85cd40, current_line = 434592720, repeating = 1, 
          getline = 0x7f7fffff6e40, cookie = 0xf1d057951b5 <process_job_event+272>}
        real_cookie = 0x0
        getline_is_func = 0
        recursive = 1
        call_depth = 1
#13 0x00000f1d057aae2a in do_cmdline_cmd (cmd=0xf1d05a79dfd "q!") at ../src/nvim/ex_docmd.c:293
No locals.
#14 0x00000f1d0584a318 in nv_Zet (cap=0x7f7fffff6f60) at ../src/nvim/normal.c:4181
No locals.
#15 0x00000f1d05842767 in normal_cmd (oap=0x7f7fffff7030, toplevel=true) at ../src/nvim/normal.c:912
        ca = {oap = 0x7f7fffff7030, prechar = 0, cmdchar = 90, nchar = 81, ncharC1 = 0, ncharC2 = 0, extra_char = 0, 
          opcount = 0, count0 = 0, count1 = 1, arg = 0, retval = 0, searchbuf = 0x0}
        c = 90
        ctrl_w = false
        old_col = 0
        need_flushbuf = false
        old_pos = {lnum = 7, col = 0, coladd = 0}
        mapped_len = 0
        idx = 91
        set_prevcount = false
        old_mapped_len = 0
        __func__ = "normal_cmd"
#16 0x00000f1d05809d0c in main_loop (cmdwin=0, noexmode=0) at ../src/nvim/main.c:747
        oa = {op_type = 0, regname = 0, motion_type = 0, motion_force = 0, use_reg_one = false, inclusive = false, 
          end_adjusted = false, start = {lnum = 0, col = 0, coladd = 0}, end = {lnum = 0, col = 0, coladd = 0}, 
          cursor_start = {lnum = 0, col = 0, coladd = 0}, line_count = 0, empty = false, is_VIsual = false, 
          block_mode = false, start_vcol = 0, end_vcol = 0, prev_opcount = 0, prev_count0 = 0}
        previous_got_int = 0
        conceal_old_cursor_line = 0
        conceal_new_cursor_line = 0
        conceal_update_lines = 0
        __func__ = "main_loop"
#17 0x00000f1d058096da in main (argc=4, argv=0x7f7fffff72a8) at ../src/nvim/main.c:534
        fname = 0xf2000134cf0 "level"
        params = {argc = 4, argv = 0x7f7fffff72a8, use_vimrc = 0x7f7fffff761b "NORC", n_commands = 0, commands = {
            0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, cmds_tofree = "\000\000\000\000\000\000\000\000\000", 
          n_pre_commands = 0, pre_commands = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, edit_type = 1, 
          tagname = 0x0, use_ef = 0x0, want_full_screen = 1, input_isatty = true, output_isatty = true, 
          err_isatty = true, headless = false, no_swap_file = 0, use_debug_break_level = -1, window_count = 1, 
          window_layout = 0, diff_mode = 0}
        reading_input = true
        __func__ = "main"
```

**Comment by fmoralesc** [comment]
Yeah, I still don't get the segfault, but I get some weird behaviour with a directory with those permissions. First, netrw lists an `../` entry (I don't get that when listing `/root`), and then there is a small pause on `:q`...

**Comment by Pyrohh** [comment]
> First, netrw lists an ../ entry (I don't get that when listing /root)

I think it's unintuitive, but as far as I can tell it's not an issue:
https://unix.stackexchange.com/questions/125972/cd-on-root-folder

Having just tested it, `cd ..` and `ls ..` while in `/` seem to be equivalent to `cd /root` and `ls /`, as mentioned in one of the answers in the SO link.

**Comment by justinmk** [comment]
Oh, found the other report of this: https://github.com/neovim/neovim/issues/3216

**Comment by fmoralesc** [comment]
@justinmk Ah, I saw that earlier today, which is why I was reminded to work on this.

**Comment by justinmk** [comment]
Looking at `event_teardown`: is it possible that `queue_process_events` is triggering a recursion into `event_teardown` which then results in `loop_close(&loop)` being called twice?

**Comment by fmoralesc** [comment]
@justinmk Not sure, but I found this notice on libuv: https://github.com/libuv/libuv/blob/v1.x/src/unix/core.c#L217-L221

File: src/nvim/path.c [hunk 2]
```diff
@@ -612,8 +612,9 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
-      || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
+  if (os_file_is_readable(*buf == NUL ? "." : (char *)buf)
+      && (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
+          || os_isdir(*buf == NUL ? (char_u *)"." : buf))) {
```
Comment (oni-link) [inline src/nvim/path.c]: `*buf == NUL ? "." : (char *)buf` could be evaluated before the `if` to make it more readable.

File: src/nvim/path.c [hunk 3]
```diff
@@ -612,8 +612,10 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
-      || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
+  char *dirpath = (*buf == NUL ? "." : (char *)buf);
+  if (os_file_is_readable(dirpath)
+      && (os_scandir(&dir, dirpath)
+          || os_isdir((char_u *)dirpath))) {
```
Comment (splinterofchaos) [inline src/nvim/path.c]: The segfault, I believed, was caused by `os_scandir()` returning `false`, freeing its internal memory, `os_isdir()` returning `true`, then `scandir_next_with_dots()` reading the invalid `Directory` object. In my testing, uv's `scandir` doesn't know the difference between an empty directory and one it doesn't have permissions to read, so making `os_scandir()` return `true` when `uv_fs_scandir()` returns `0` and not free its memory should fix the segfault. (`uv_fs_scandir()` returns zero when the directory is empty, and when it doesn't have permissions--in either case, `os_scandir_next()` will be safe to call.)

The `os_isdir()` call is a serious problem. Please modify `os_scandir()` to fix this bug and remove the isdir check. The `os_file_is_readable()` check is good, though, so we don't return `"."` and `".."` on non-readable directories (which is consistent with vanilla vim).

Comment (splinterofchaos) [inline src/nvim/path.c]: Thought I should mention: I did write up and test the solution I recommended, but I've been waiting about a week to get permission from my employer to send it.

Comment (justinmk) [inline src/nvim/path.c]: Looking forward to it!

Comment (fmoralesc) [inline src/nvim/path.c]: Thanks for taking a look, @splinterofchaos, really appreciated. 

I wonder why I haven't had a single segfault when testing this, then...

**Comment by fmoralesc** [comment]
Updated this as per @splinterofchaos suggestions. I think I got it right, but I would appreciate comments.

**Comment by splinterofchaos** [comment]
LGTM :+1: 

Tested locally and it works as expected.

**Comment by splinterofchaos** [comment]
I think you should be able to add  #3221 to the bugs this fixes.

**Comment by fmoralesc** [comment]
Rebased. Thanks for the help you all, I'm mentioning it in the commit.

**Comment by Pyrohh** [comment]
I'm still hitting the assertion I mentioned above, anyone else? This is on any directory, regardless of permission.

File: src/nvim/path.c [hunk 4]
```diff
@@ -612,8 +612,9 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
-      || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
+  char *dirpath = (*buf == NUL ? "." : (char *)buf);
+  if (os_file_is_readable(dirpath)
+      && os_scandir(&dir, dirpath)) {
```
Comment (oni-link) [inline src/nvim/path.c]: Fits in one line.

Comment (fmoralesc) [inline src/nvim/path.c]: :+1:

**Comment by oni-link** [comment]
@Pyrohh, cannot reproduce your assertion on gentoo.

**Comment by justinmk** [comment]
I believe BSD tends to find bugs more quickly because its enables allocator "poisoning" features by default.

**Comment by Pyrohh** [comment]
@justinmk I'm glad you mentioned that, as I didn't say I'm building without Jemalloc, although I wonder if that's actually related. Perhaps I'll open an issue at the libuv repo.

**Comment by justinmk** [comment]
@Pyrohh Disabling jemalloc has uncovered problems in another case: https://github.com/neovim/neovim/issues/3150#issuecomment-131240462

Also, MSan reports a problem with `os_stat` which @fwalch and I tried to dig into here: https://github.com/neovim/neovim/pull/2807#discussion-diff-32729362 But we did not reach a conclusion.

**Comment by fmoralesc** [comment]
@Pyrohh can you test this PR over #3289? jemalloc 4.0 supposedly adds support for OpenBSD and DragonflyBSD, so disabling jemalloc shouldn't be necessary anymore.

**Comment by fmoralesc** [comment]
Just tested on Linux without jemalloc and I still can't reproduce the segfault.

**Comment by justinmk** [comment]
But again, Linux (in)famously memory access bugs, whereas BSD typically does not. Unfortunately I can't find a reference other than my previous comment: https://github.com/neovim/neovim/issues/1375#issuecomment-62852376

**Comment by justinmk** [comment]
You could try compiling with the malloc options enabled: http://man7.org/linux/man-pages/man3/mallopt.3.html

**Comment by fmoralesc** [comment]
Building with `MALLOC_CHECK_` doesn't seem to make any difference. Let's wait on what @Pyrohh tells us about building with jemalloc 4.0.

**Comment by Pyrohh** [comment]
> This may be irrelevant since the issue reported by @Pyrohh is not a segfault, but a failed assertion.

Indeed, but there's no harm in trying out the newer Jemalloc release (will report back shortly).

**Comment by Pyrohh** [comment]
Huh, `ktrace` shows that `sched_yield()` is being called repeatedly, which only stops when I Ctrl-c. This is way out of my depth :grin:

**Comment by Pyrohh** [comment]
Sorry if I'm spamming your inbox (this is the last one in my current series of posts) but I forgot to mention that the `sched_yield()` thing is happening using Jemalloc 4.0.0.

**Comment by justinmk** [comment]
Merged, thanks all.

**Comment by justinmk** [comment]
>  I'll open an issue for tracking the assert, if there isn't one already.

@Pyrohh https://github.com/neovim/neovim/issues/3216

**Comment by fmoralesc** [comment]
Yes, thanks everyone who helped with this.

**Comment by Pyrohh** [comment]
Thanks @justinmk.

**Comment by tarruda** [comment]
> I noticed that we never close fs_loop in os/fs.c. @tarruda is there any chance that could cause an issue?

@justinmk no, according to @Pyrohh's backtrace, the abort happens in the main event loop.

@Pyrohh can you provide some more information about that core file? I'd like to see the values of the `handle` variable in frame 3. In gdb type:

```
frame 3
p *handle
```

**Comment by Pyrohh** [comment]
@tarruda this is bizarre, I can't reproduce it anymore. Could #3246 have done something about it?

I semi-frequently update OpenBSD -current, so perhaps it was a libc bug, although that's merely speculation...

**Comment by dnmario** [comment]
Dear Michael Reed,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [RFC] path: check if directories are readable when"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

**Pull request closed without merging**